File: algorithms/257.Binary_Tree_Paths/Binary_Tree_Paths.cc
 class Solution {
    public:

            vector<string> binaryTreePaths(TreeNode* root) {
                vector<string> v;
                if(root)
                    preorder(v,root,"");
                return v;
            }
            void preorder(vector<string>& v,TreeNode* r,string t){
                if(!r)
                    return;
                if(!t.empty())
                    t+=("->"+to_string(r->val));
                else t+=to_string(r->val);
                if(r->left||r->right){
                    preorder(v,r->left,t);
                    preorder(v,r->right,t);
                }else{
                    v.push_back(t);
                }
            }
        };
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void binaryTreePaths(vector<string>& result, TreeNode* root, string t) {
    if(!root->left && !root->right) {
        result.push_back(t);
        return;
    }

    if(root->left) binaryTreePaths(result, root->left, t + "->" + to_string(root->left->val));
    if(root->right) binaryTreePaths(result, root->right, t + "->" + to_string(root->right->val));
}

vector<string> binaryTreePaths(TreeNode* root) {
    vector<string> result;
    if(!root) return result;

    binaryTreePaths(result, root, to_string(root->val));
    return result;
}
};
// Binary Tree Paths
class Solution {
  vector<string> r;
  void f(TreeNode *x, string s) {
    if (! x) return;
    s = s.empty() ? to_string(x->val) : s+"->"+to_string(x->val);
    f(x->left, s);
    f(x->right, s);
    if (! x->left && ! x->right)
      r.push_back(s);
  }
public:
  vector<string> binaryTreePaths(TreeNode *root) {
    f(root, "");
    return r;
  }
};

File: algorithms/257.Binary_Tree_Paths/Binary_Tree_Paths.java
<<<<<<< HEAD
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
   public List<String> binaryTreePaths(TreeNode root) {
    List<String> answer = new ArrayList<String>();
    if (root != null) searchBT(root, "", answer);
    return answer;
}
private void searchBT(TreeNode root, String path, List<String> answer) {
    if (root.left == null && root.right == null) answer.add(path + root.val);
    if (root.left != null) searchBT(root.left, path + root.val + "->", answer);
    if (root.right != null) searchBT(root.right, path + root.val + "->", answer);
}
}
public class Solution {
 public List<String> binaryTreePaths(TreeNode root) {

        List<String> paths = new LinkedList<>();

        if(root == null) return paths;

        if(root.left == null && root.right == null){
            paths.add(root.val+"");
            return paths;
        }

         for (String path : binaryTreePaths(root.left)) {
             paths.add(root.val + "->" + path);
         }

         for (String path : binaryTreePaths(root.right)) {
             paths.add(root.val + "->" + path);
         }

         return paths;

    }
}
public class Solution {
    public List<String> binaryTreePaths(TreeNode root) {
        List<String> l=new ArrayList<>();
        if(root!=null)
            pre(l,root,"");
        return l;
    }
     void pre(List<String> l,TreeNode r,String s){
         if(r==null)return;
         if(s.isEmpty())
            s+=r.val;
        else s+=("->"+r.val);
        if(r.left!=null||r.right!=null){
            pre(l,r.left,s);
            pre(l,r.right,s);
        }else
            l.add(s);
     }
=======
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
   public List<String> binaryTreePaths(TreeNode root) {
    List<String> answer = new ArrayList<String>();
    if (root != null) searchBT(root, "", answer);
    return answer;
}
private void searchBT(TreeNode root, String path, List<String> answer) {
    if (root.left == null && root.right == null) answer.add(path + root.val);
    if (root.left != null) searchBT(root.left, path + root.val + "->", answer);
    if (root.right != null) searchBT(root.right, path + root.val + "->", answer);
}
}
public class Solution {
 public List<String> binaryTreePaths(TreeNode root) {

        List<String> paths = new LinkedList<>();

        if(root == null) return paths;

        if(root.left == null && root.right == null){
            paths.add(root.val+"");
            return paths;
        }

         for (String path : binaryTreePaths(root.left)) {
             paths.add(root.val + "->" + path);
         }

         for (String path : binaryTreePaths(root.right)) {
             paths.add(root.val + "->" + path);
         }

         return paths;

    }
}
public class Solution {
    public List<String> binaryTreePaths(TreeNode root) {
        List<String> l=new ArrayList<>();
        if(root!=null)
            pre(l,root,"");
        return l;
    }
     void pre(List<String> l,TreeNode r,String s){
         if(r==null)return;
         if(s.isEmpty())
            s+=r.val;
        else s+=("->"+r.val);
        if(r.left!=null||r.right!=null){
            pre(l,r.left,s);
            pre(l,r.right,s);
        }else
            l.add(s);
     }
>>>>>>> 6200c8704614e918c8bfa5357c648dd1b4f7eb74
}